packages/story-editor/src/components/floatingMenu/elements/shared/focusTrapButton.js

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+
+import PropTypes from 'prop-types';
+import styled, { css } from 'styled-components';
+import { Button, BUTTON_TYPES } from '@googleforcreators/design-system';
+import { forwardRef } from '@googleforcreators/react';
+import { __, sprintf } from '@googleforcreators/i18n';
+import { v4 as uuidv4 } from 'uuid';
+
+const _FocusTrapButton = styled(Button).attrs({ type: BUTTON_TYPES.DEFAULT })`
+  background-color: transparent;
+  display: flex;
+  align-items: center;
+  justify-content: flex-start;
+  outline: none;
+  margin: 0;
+  padding: 0;
+  min-width: inherit;
+  ${({ styleOverride }) =>
+    css`
+      ${styleOverride}
+    `}
+`;
+_FocusTrapButton.propTypes = {
+  styleOverride: PropTypes.array,
+};
+
+export const handleReturnTrappedFocus = (e, buttonRef) => {
+  // only bubble up for moving focus
+  if (e.key === 'Tab') {
+    e.preventDefault();
+    buttonRef.current.focus();
+  } else {
+    e.stopPropagation();
+  }
+};
+
+/**
+ * Intended as a last resort effort to wrangle nested inputs within a <ContextMenu>. Freeform inputs should be avoided within menus if at all possible. Because this is a last resort, it is a shared element in the floating menu not the design-system.
+ * When it's necessary, wrap the freeform input in the <FocusTrapButton> and pass the associated handleReturnTrappedFocus as onKeyDown of the designated input.
+ * You will also need to give the input the imported CONTEXT_MENU_SKIP_ELEMENT class from design system. This tells the context menu to ignore the input in its focusableChildren so it relies on the FocusTrapButton as the focusable element for that input. This allows the freeform inputs to be typed into and arrowed around and then tab to get back into the menu.
+ *
+ * **Sample Usage**
+ * ```js
+ *
+ * function SomeComponent() {
+ * const buttonRef = useRef();
+ * const inputRef = useRef();
+ *
+ * <FocusTrapButton ref={buttonRef} inputRef={inputRef} inputLabel='Font size'>
+ * <input ref={inputRef} aria-label="Font size" onKeyDown={(e) => handleReturnTrappedFocus(e, buttonRef)} inputClassName={CONTEXT_MENU_SKIP_ELEMENT} />
+ * </FocusTrapButton>
```

Comment:
I'm cool with this implementation. This is isolated and currently working so there's no need to change this.

but if you do run into a scenario like this again an alternative API consideration could be to grab the input with a query selector instead of passing in a ref prop, then mutating that DOM node directly. something like:
```js
function FloatingMenuInputKeyboardNavWrapper(props) {
  const callbackRef = (node) => {
    // maybe need some cleanup. idk
    if(!node) {
      return;
    }
    
    const childInputNode = node.querySelector('> input');
    if (!childInputNode) {
      throw new Error('Direct child of FloatingMenuInputKeyboardNavWrapper must be an input')
    }
    
    // childInputNode manipulations
    // - add CONTEXT_MENU_SKIP_ELEMENT class to childInputNode
    // - add keydown handlers between node & childInputNode to change focus
    // - ensure any tabIndex stuffs
  }
  
  return <Button ref={callbackRef} {...props} />
}
```
Could then just wrap inputs with this without having to alter them. Anyway, I'm sure this approach would introduce some edge case regressions and you already have the current approach all ironed out. Just mentioning cus I see how this is a pickle, and if you ever run into this pickle in the future, this is an alternative API design that could be a lil more self contained.

Reply:
welp ignore all of this 😅 . Integrating this with the color input would have been a pain with this approach I think. didn't get to that till after I wrote this comment

Reply:
yeah the color input was a really bummer curveball :(